README.md
cacti-gpsmaps
=============

Originally developped by Andy Aloia, GPSMaps is a plugin that provides an integration of Google Maps with the Cacti network graphing solution.


Docs/changelog.txt
Changelog:
Version 1.0
-Display map with hard coded static parameters.
-Enter Hosts Lat & Long in host settings

Version 1.01
-Ability to set parameters for map

Version 1.02
- few optimizations and less DB calls

Version 1.1
-added preliminary IP range tracking
-cleaned up some of the code for settings and such

Version 1.11
-Cleaned up code for the IP tracking
-Added multiple colors for points based on type of device(currently only hard coded)
-Added information when point is clicked

Version 1.2
-Allowed ways for users to add custom GPS types (non-GUI)
-set width to be a permenant 100% (still have not fixed height)
-In settings changed tab name to GPS Map
-Added alot of in-code commenting.

Version 1.3
6/11/2009

-Implemented PIA 2.x standards
-Fixed partially, Bug when performing an IP analysis and no IP address is specified in hostname area. Currently it will not display those devices.
-Fixed some minor inconsistancies involing the XML output to google maps.
-Ensured that install/uninstall works and removed database columns

Version 1.31
-Alot of changes in preparation for the addition of circle overlays for coverage. 
-Created a base to work from

Version 1.32
-Made for the addition of static circles on simple points based on lat/long
-Started work on how to get distance from tower bases. 
-most of the PHP for the coverage XML has been written

Version 1.33
-Found some math errors and corrected
-First dynamic plotting of overlay done. 
-Modified the coverage XML to make each tower an object and treat them as such. 
-(speculation on the use of objects for the IP analysis)

Version 1.34
-Allowed for towers in the same IP range to graph a circle. This might not be the best way to approach this issue.
But seeing that there is no way to get actual connected subscribers, IP range is the only way to do this.

Version 1.35
-Increated float range in database for increased accuracy. Up to 13 characters for Lat/Long now
-removed the integer conversions in code to increase accurrcy
-Added customizable options in the settings.

Version 1.36
-Fixed bug that mistakenly displayed incorrect towers and nodes when in regular mode.
-Fixed bug that displayed all towers no matter what IP range settings when in coverage mode.
-Changed the way the IP Region adds to an array. This is to make it a dynamic array and decrease wasted space.

Version 1.37
-Fixed bug that would have multiple overlays for one tower, each overlay being a different size.
-Fixed bug that would keep towers from overlaying proper distance.
-Fixed When multiple points were on the same IP range they could potentially have a circle drawn about them, if conditions were right.
-Added multiple AP Overlay mode. This is to allow multiple overlays per IP range. By default it only allows one overlay per IP range. (Ip range is defined as the 192.168.XXX.YYY Where XXX differs between ranges)
-Added link to the host graphs from the icon pop-up.
-Will only display ENABLED hosts. Will add the option later to display all hosts.

Version 1.4
-Complete overhaul of the Host/Tower management. Moved over to an object oriented design that effectively reduced the amount of database calls at minimum of half. This should speed up the displaying of the map.
-The top grey bar now dynamically sizes to the height of the text.
-The final IP analysis area displays all IP address for that range, this links to that hosts graphs. This can make a problem of having multiple rows of grey bars with IP addresses. Looking for solutions.
-A maximum of 6 IP address will be stacked before moving onto the next column. 
-In the GPSMap settings, I have made simpler the setup of the tower Host ID. Just select the template that is being used by your towers/APs.
-New Towers Only mode selectable in the settings, guess what it does? Thats right, displays on the towers. It can also display coverage maps
-Added a print button. DOES NOT PRINT OVERLAY. Takes a little bit of guess work to get a pic of what you want, but at least it works.
-Made map navigation to a full nav bar. It got annoying just using the (+) and (-) buttons.

Version 1.41
-Added the UP/Down/Recovery viewing. Hosts change color based on status. Colors correspond to cacti designated colors.
-Added the ability to display disabled hosts. These wil show up as BLACK markers to ensure that a difference can be seen.
-Added a preliminary test for overlaying current radar information. Take a look at the Weather.php file for more info.
-Modified the navigation to allow for mouse wheel usage.

*Yeah... forgot to update the changelog for all of the changes. So v1.5 because it is a major update
Version 1.5
-Weather has been disabled. Some code is still left in there for the prying people
-KML exportin is fully supported
-Templates for up/down/recover have been set
-GPS maps outputs the templates and set colors and images accordingly.
-I havent fully fixed the setup database script yet, so...there will need some work
-Added a few parameters to the host class to accomodate the changes
-So long as original server accepts connections Google earth should get custom images.
-Still small bug with the console always being red
-Removed the custom hosts file, it is no longer needed.
-Several new files, to accomodate GUI updates
-A big thanks to the thold plugin for helping me out.

Version 1.51

-Minor bugfixes, there was one leftover <php= VAR ?> left around
-some small tweaks to the KMLcreation file to prevent errors on 0 indexes.
-small fix to remove some pulled settings.

Version 1.62
-Changed the processing of points. 
-Reduced some DB queries by half.
-Thanks to Nate for the tabs
http://groups.google.com/group/Google-Maps-API/browse_thread/thread/b023708935be9392/ae9ca67ef4e8b136%3Fshow_docid=ae9ca67ef4e8b136
-Nodes close to each other are grouped into tabs. There is a threashold setting for this. 

print.php
<?php
/*
 +-------------------------------------------------------------------------+
 | Copyright (C) 2009-2013 Andrew Aloia                                    |
 | Copyright (C) 2014 Wixiweb                                              | 
 |                                                                         |
 | This program is free software; you can redistribute it and/or           |
 | modify it under the terms of the GNU General Public License             |
 | as published by the Free Software Foundation; either version 2          |
 | of the License, or (at your option) any later version.                  |
 |                                                                         |
 | This program is distributed in the hope that it will be useful,         |
 | but WITHOUT ANY WARRANTY; without even the implied warranty of          |
 | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           |
 | GNU General Public License for more details.                            |
 +-------------------------------------------------------------------------+
 | http://www.cacti.net/                                                   |
 +-------------------------------------------------------------------------+
*/
?>
<script language="javascript">
document.write(window.opener.document.getElementById("map").innerHTML);
window.print();
</script>

js/GPSMaps.js
/*
 +-------------------------------------------------------------------------+
 | Copyright (C) 2009-2013 Andrew Aloia                                    |
 | Copyright (C) 2014 Wixiweb                                              | 
 |                                                                         |
 | This program is free software; you can redistribute it and/or           |
 | modify it under the terms of the GNU General Public License             |
 | as published by the Free Software Foundation; either version 2          |
 | of the License, or (at your option) any later version.                  |
 |                                                                         |
 | This program is distributed in the hope that it will be useful,         |
 | but WITHOUT ANY WARRANTY; without even the implied warranty of          |
 | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           |
 | GNU General Public License for more details.                            |
 +-------------------------------------------------------------------------+
 | http://www.cacti.net/                                                   |
 +-------------------------------------------------------------------------+
*/

var gpsmap = {

    map : null,
    refreshMap : '',
    initialLat : 0,
    initialLng : 0,
    initialZoom : 0,
    liColor : '',
    liWidth : '',
    liOpa : '',
    fillColor : '',
    fillOpa : '',
    circleQuality : '',
    enableWeather : '',
    coverageOverlay : '',
    downloadURL : '',
    markerArray : [],
    APArray : [],
    t_error : 0,

    // -------------------------------------------------------------------------
    loader : function() {  
        this.resize();
        this.gload();
        if (this.refreshMap !== '0') {
            setInterval(function() {gpsmap.gload();}, (this.refreshMap * 60000));
        }
    },
            
    // -------------------------------------------------------------------------
    gload : function() {
        var self = this;

        this.map = new google.maps.Map(
            document.getElementById('map'),
            {
                center: new google.maps.LatLng(self.initialLat, self.initialLng),
                zoom: self.initialZoom,
                mapTypeId: google.maps.MapTypeId.ROADMAP
            }
        );

        this.loadMarkersData(this.downloadURL, function(data) {
            var xml = self.parseMarkersDataToXml(data),
                markers = xml.documentElement.getElementsByTagName('marker');
            ;

            self.putMarkersOnMap(markers);

            // if radius is set, we need to draw a circle.
            if (self.coverageOverlay) {
                for (var i = 0; i < self.APArray.length; i++) {
                    self.drawCoverage(self.APArray[i]);
                }
            }
        });
    },

    // -------------------------------------------------------------------------
    drawCoverage : function(xmlMarker) {
        var center = new google.maps.LatLng(
            parseFloat(xmlMarker.getAttribute('lat')),
            parseFloat(xmlMarker.getAttribute('lng'))
        );

        var radius = xmlMarker.getAttribute('radius');
        var start = parseInt(xmlMarker.getAttribute('start'));
        var stop = parseInt(xmlMarker.getAttribute('stop'));
        var points = [];
        var degrees = stop - start;

        // calculating km/degree
        var latConv = google.maps.geometry.spherical.computeDistanceBetween(
            center,
            new google.maps.LatLng(center.lat() + 0.1, center.lng())
        ) / 100;
        var lngConv = google.maps.geometry.spherical.computeDistanceBetween(
            center,
            new google.maps.LatLng(center.lat(), center.lng() + 0.1)
        ) / 100;

        // loop 
        if (degrees > 360) {
            degrees = 360;
        }

        var step = parseInt(degrees / this.circleQuality);
        if (step === 0) {
            step = 1;
        }

        // if not a circle
        if (degrees < 360) {
            // opens the area
            points.push(center);
        }

        for (var i = start; i < stop; i += step) {
            points.push(
                new google.maps.LatLng(
                    center.lat() + (radius / latConv * Math.cos(i * Math.PI / 180)),
                    center.lng() + (radius / lngConv * Math.sin(i * Math.PI / 180))
                )
            );
        }

        if (degrees === 360) {
            // closes the circle
            points.push(points[0]); 
        }
        else {
            // closes the area
            points.push(center);
        }

        new google.maps.Polygon({
            clickable: false,
            fillColor : '#' + this.fillColor,
            fillOpacity : this.fillOpa,
            map : this.map,
            paths : points,
            strokeColor : '#' + this.liColor,
            strokeOpacity : this.liOpa,
            strokeWeight : this.liWidth
        });
    },

    // -------------------------------------------------------------------------
    setOuterBounds : function(num) {
        var marker = this.markerArray[num],
            n_outer = marker.startcoord.lat() + .5 * ((marker.n_outer - marker.s_outer + .00075) / Math.pow(2, -(17 - this.map.getZoom()))),
            s_outer = marker.startcoord.lat() - .5 * ((marker.n_outer - marker.s_outer) / Math.pow(2, -(17 - this.map.getZoom()))),
            e_outer = marker.startcoord.lng() + .5 * ((marker.e_outer - marker.w_outer + .00025) / Math.pow(2, -(17 - this.map.getZoom()))),
            w_outer = marker.startcoord.lng() - .5 * ((marker.e_outer - marker.w_outer + .00025) / Math.pow(2, -(17 - this.map.getZoom())))
        ; 

        return new google.maps.LatLngBounds(
            new google.maps.LatLng(s_outer, w_outer),
            new google.maps.LatLng(n_outer, e_outer)
        );
    },

    // -------------------------------------------------------------------------
    createMarker : function(xmlMarker) {
        var self = this,
            name = xmlMarker.getAttribute('name'),
            status = xmlMarker.getAttribute('status'),
            point = new google.maps.LatLng(
                parseFloat(xmlMarker.getAttribute('lat')),
                parseFloat(xmlMarker.getAttribute('lng'))
            ),
            joined = xmlMarker.getAttribute('templateId') + status,
            mapMarker = new google.maps.Marker({
                position: point,
                icon: (self.customIcons[joined] == undefined)
                    ? self.customIcons[status]
                    : self.customIcons[joined]
                ,
                title: name
            })
        ;

        mapMarker.startimport = 1;
        mapMarker.importance = 1;
        mapMarker.targeted = 0;
        mapMarker.startcoord = point;
        mapMarker.line = null;
        mapMarker.shifted = 0;
        mapMarker.name = name;
        mapMarker.html = '<a href="../../graph_view.php?action=preview&host_id='
                       + xmlMarker.getAttribute('id')
                       + '"><b>'
                       + name 
                       + '</b></a><br />'
                       + xmlMarker.getAttribute('address')
                       + '<br /> Avalability: '
                       + xmlMarker.getAttribute('availability')
                       + '<br /> Type: '
                       + xmlMarker.getAttribute('type')
                       + '<br /> Latency: '
                       + xmlMarker.getAttribute('latency')
        ;

        var shift = this.t_error + .0002,
            n_coord = mapMarker.getPosition().lat() + this.t_error,
            s_coord = mapMarker.getPosition().lat() - this.t_error,
            e_coord = mapMarker.getPosition().lng() + this.t_error,
            w_coord = mapMarker.getPosition().lng() - this.t_error
        ;

        mapMarker.inBounds = new google.maps.LatLngBounds(
            new google.maps.LatLng(s_coord, w_coord),
            new google.maps.LatLng(n_coord, e_coord)
        );

        mapMarker.n_outer = mapMarker.startcoord.lat() + shift;
        mapMarker.s_outer = mapMarker.startcoord.lat() - shift;
        mapMarker.e_outer = mapMarker.startcoord.lng() + shift;
        mapMarker.w_outer = mapMarker.startcoord.lng() - shift;
        mapMarker.outBounds = null;

        return mapMarker;
    },

    // -------------------------------------------------------------------------
    resize : function() {
        var frame = document.getElementById('map'),
            htmlheight = document.body.parentNode.scrollHeight,
            windowheight
        ;                     

        if (window.innerHeight) {
            windowheight = window.innerHeight;
        }
        else if (document.documentElement && document.documentElement.clientHeight) {
            windowheight = document.documentElement.clientHeight;
        }
        else if (document.body) {
            windowheight = document.body.clientHeight;
        }

        if (htmlheight < windowheight) {
            document.body.style.height = windowheight + 'px';
            frame.style.height = parseInt(windowheight - 105) + 'px';
        }
        else {
            document.body.style.height = htmlheight + 'px';
            frame.style.height = parseInt(htmlheight - 105) + 'px'; 
        }
    },

    // -------------------------------------------------------------------------
    loadMarkersData : function(url, callback) {
        var xhr = (window.XMLHttpRequest)
                ? new XMLHttpRequest()
                : new ActiveXObject('Microsoft.XMLHTTP')
            ,
            onError = function(httpStatus, httpText) {
                alert(
                    'A network error has occured while retrieving markers data : '
                    + httpStatus
                    + ' - '
                    + httpText
                );
            }
        ;

        if (xhr.addEventListener) {
            xhr.addEventListener('load', function(e) {
                if (e.target.status !== 200 && e.target.status !== 304 && e.target.status !== 0) {
                    onError(e.target.status, e.target.statusText);
                }
                callback(e.target.responseText);
            }, false);
            xhr.addEventListener('error', function(e) {
                onError(e.target.status, 'error');
            }, false);
        } else if (el.attachEvent)  {
            xhr.attachEvent('load', function(e) {
                if (e.target.status !== 200 && e.target.status !== 304 && e.target.status !== 0) {
                    onError(e.target.status, e.target.statusText);
                }
                callback(e.target.responseText);
            }, false);
            xhr.attachEvent('error', function(e) {
                onError(e.target.status, 'error');
            }, false);
        }

        xhr.open('GET', url, true);
        xhr.setRequestHeader('X_REQUESTED_WITH', 'XMLHttpRequest');
        xhr.send();
    },

    // -------------------------------------------------------------------------
    parseMarkersDataToXml : function(xmlText) {
        // https://gist.github.com/stevenaw/1305672
        try {
            var xml = null;
            if (window.DOMParser) {
                var parser = new DOMParser();
                xml = parser.parseFromString(xmlText, 'text/xml');
                var found = xml.getElementsByTagName('parsererror');

                if (!found || !found.length || !found[ 0 ].childNodes.length) {
                    return xml;
                }

                return null;
            }
            else {
                xml = new ActiveXObject('Microsoft.XMLDOM');
                xml.async = false;
                xml.loadXML(xmlText);

                return xml;
            }
        }
        catch (e) {
            return null;
        }
    },

    //-------------------------------------------------------------------------- 
    putMarkersOnMap : function(markers) {
        var infoWindow = new google.maps.InfoWindow({}),
            infoBubble = new InfoBubble({})
        ;

        for (var i = 0; i < markers.length; i++) {
            var self = this,
                mapMarker = this.createMarker(markers[i])
            ;

            mapMarker.setMap(this.map);

            this.markerArray.push(mapMarker);
            if (markers[i].getAttribute('radius') > 0) {
                this.APArray.push(markers[i]);
            }

            google.maps.event.addListener(mapMarker, 'click', function(e) {
                var step = 0;

                infoWindow.close();
                infoBubble.close();
                while (infoBubble.tabs_.length) {
                    infoBubble.removeTab(infoBubble.tabs_.length - 1);
                }

                for (var x = 0; x < self.markerArray.length; x++) {
                    if (this.inBounds.contains(self.markerArray[x].getPosition()) === true
                        && self.markerArray[x].getVisible() === true
                    ) {
                        step++;
                        self.markerArray[x].targeted = 1;
                        self.markerArray[x].outBounds = self.setOuterBounds(x);
                    }
                }

                if (step === 1) {
                    this.targeted = 0;
                    this.outBounds = null;

                    infoWindow.setContent(this.html);
                    infoWindow.open(self.map, this);
                }
                else if (step > 1) {
                    var tempCount = 0;
                    for (var x = 0; x < self.markerArray.length; x++) {
                        if (self.markerArray[x].targeted === 1) {
                            tempCount++;
                            infoBubble.addTab(
                                'Point ' + tempCount,
                                '<div style="width:' + (88 * step) + 'px;"></div>' + self.markerArray[x].html
                            );
                        }
                        self.markerArray[x].targeted = 0;
                        self.markerArray[x].outBounds = null;
                    }
                    infoBubble.open(self.map, this);
                }
            });
        }
    }
};

class/hosts_class.php
<?php
/*
 +-------------------------------------------------------------------------+
 | Copyright (C) 2009-2013 Andrew Aloia                                    |
 | Copyright (C) 2014 Wixiweb                                              | 
 |                                                                         |
 | This program is free software; you can redistribute it and/or           |
 | modify it under the terms of the GNU General Public License             |
 | as published by the Free Software Foundation; either version 2          |
 | of the License, or (at your option) any later version.                  |
 |                                                                         |
 | This program is distributed in the hope that it will be useful,         |
 | but WITHOUT ANY WARRANTY; without even the implied warranty of          |
 | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           |
 | GNU General Public License for more details.                            |
 +-------------------------------------------------------------------------+
 | http://www.cacti.net/                                                   |
 +-------------------------------------------------------------------------+
*/

class host{
	var $id = "0";
	var $type = "0";
	var $lat = "0";
	var $long = "0";
	var $iprange = "";
	var $description = "";
	var $hostname = "";
	var $radius = "0";
	var $avail = "";
	var $status = "";
	var $showMap = "1";
	var $latency = "0";
	var $coverage = "1";
	var $upimage = "";
	var $downimage = "";
	var $recoverimage = "";
	var $start = "0";
	var $stop = "360";
	var $group = "0";

function host($id,$type,$lat,$long,$iprange,$description,$hostname,$radius,$avail,$status,$latency,$coverage,$upimage,$downimage,$recoverimage,$start,$stop,$group){
	
	$this->id = $id;
	$this->type = $type;
	$this->lat = $lat;
	$this->long = $long;
	$this->iprange = $iprange;
	$this->description = $description;
	$this->hostname = $hostname;
	$this->radius = $radius;
	$this->avail = $avail;
	$this->status = $status;
	$this->latency = $latency;
	if($coverage === "on")
		$this->coverage = 1;
	else
		$this->coverage = 0;
	$this->upimage = $upimage;
	$this->downimage = $downimage;
	$this->recoverimage = $recoverimage;
	$this->start = $start;
	$this->stop = $stop;
	$this->group = $group;

}
}
?>


setup.php
<?php
/*
 +-------------------------------------------------------------------------+
 | Copyright (C) 2009-2013 Andrew Aloia                                    |
 | Copyright (C) 2014 Wixiweb                                              | 
 |                                                                         |
 | This program is free software; you can redistribute it and/or           |
 | modify it under the terms of the GNU General Public License             |
 | as published by the Free Software Foundation; either version 2          |
 | of the License, or (at your option) any later version.                  |
 |                                                                         |
 | This program is distributed in the hope that it will be useful,         |
 | but WITHOUT ANY WARRANTY; without even the implied warranty of          |
 | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           |
 | GNU General Public License for more details.                            |
 +-------------------------------------------------------------------------+
 | http://www.cacti.net/                                                   |
 +-------------------------------------------------------------------------+
*/

function plugin_gpsmap_install () {
	global $config;
	api_plugin_register_hook('gpsmap', 'top_header_tabs',       'gpsmap_show_tab',             'includes/setup/tabs.php');
	api_plugin_register_hook('gpsmap', 'top_graph_header_tabs', 'gpsmap_show_tab',             'includes/setup/tabs.php');
	api_plugin_register_hook('gpsmap', 'config_arrays',         'gpsmap_config_arrays',        'includes/setup/settings.php');
	api_plugin_register_hook('gpsmap', 'config_settings',       'gpsmap_config_settings',      'includes/setup/settings.php');
	api_plugin_register_hook('gpsmap', 'draw_navigation_text',  'gpsmap_draw_navigation_text', 'includes/setup/settings.php');
	api_plugin_register_hook('gpsmap', 'api_device_save',       'gpsmap_api_device_save', 	'includes/setup/settings.php');
	api_plugin_register_hook('gpsmap', 'config_form', 	    'gpsmap_config_form',          'setup.php');
	api_plugin_register_hook('gpsmap', 'poller_bottom', 	    'gpsmap_poller_bottom', 	'includes/polling.php');
	
	api_plugin_register_realm('gpsmap',	'gpstemplates.php,gpstemplates_add.php','Configure GPSMap', 1);
	api_plugin_register_realm('gpsmap',	'gpsmap.php', 'View GPSMap', 1);

	include_once($config['base_path'] . '/plugins/gpsmap/includes/setup/database.php');
	gpsmap_setup_database();
}


function plugin_gpsmap_uninstall () {
	//gpsmap_remove_database();
}        
function plugin_gpsmap_check_config () {
	//Here we will check to ensure everything is configured 
	gpsmap_check_upgrade ();
	return true;
}

function plugin_gpsmap_upgrade () {
	// Here we will upgrade to the newest version
	gpsmap_check_upgrade ();
	return false;
} 
 
function plugin_gpsmap_version () {
	return gpsmap_version();
}

function gpsmap_check_upgrade () {
	global $config;
	$files = array('gpsmap.php','gpstemplates.php','poller.php');
	if (isset($_SERVER['PHP_SELF']) && !in_array(basename($_SERVER['PHP_SELF']), $files))
		return;
	$current = plugin_gpsmap_version();
	$current = $current['version'];
	$old = read_config_option('plugin_gpsmap_version', TRUE);
	if ($current != $old) {
		include_once($config['base_path'] . '/plugins/gpsmap/includes/database.php');
		gpsmap_upgrade_database ();
	}
}

function gpsmap_version () {
    return array(
        'name'      => 'gpsmap',
        'version'   => '1.8.0',
        'longname'  => 'GPS Mapping',
        'author'    => 'Andy Aloia / Wixiweb',
        'homepage'  => 'http://www.wixiweb.com/',
        'email'     => 'contact@wixiweb.fr',
        'url'       => 'https://github.com/wixiweb/cacti-gpsmaps'
    );
}

//defines latitude and longitude for devices
function gpsmap_config_form () {
	global $fields_host_edit;
	$fields_host_edit2 = $fields_host_edit;
	$fields_host_edit3 = array();
	foreach ($fields_host_edit2 as $f => $a) {
		$fields_host_edit3[$f] = $a;
		if ($f == 'disabled') {
			$fields_host_edit3['gpsSpacer'] = array(
            			'friendly_name' => 'GPS Settings',
			       'method' => 'spacer',
		            );
			$fields_host_edit3['GPScoverage'] = array(
				'friendly_name' => 'Overlay Inclusion',
				'description' => 'Disable to plot host only, not included in coverage overlay.',
				'method' => 'checkbox',
				'value' => '|arg1:GPScoverage|',
				'default' => 'on',
		        );

			$fields_host_edit3['latitude'] = array(
				'friendly_name' => 'Latitude',
				'description' => 'The devices latitude coordinates',
				'method' => 'textbox',
				'max_length' => 13,
				'value' => '|arg1:latitude|',
				'default' => '',
			);
			$fields_host_edit3['longitude'] = array(
				'friendly_name' => 'Longitude',
				'description' => 'The devices longitude coordinates',
				'method' => 'textbox',
				'max_length' => 13,
				'value' => '|arg1:longitude|',
				'default' => '',
			);
			//Only apply this rule if on a Host page with an ID number present and is an AP device
			global $url_path;
			if ((isset ($_GET["id"]))&&($_SERVER["PHP_SELF"] == $url_path."host.php")){
				$did = $_GET["id"];
				//$result = mysql_query("SELECT AP FROM `host` RIGHT JOIN gpsmap_templates ON host.host_template_id = gpsmap_templates.templateID WHERE id=".$did."");
				//$row = mysql_fetch_array($result,MYSQL_ASSOC);
				$row =db_fetch_assoc("SELECT AP FROM `host` RIGHT JOIN gpsmap_templates ON host.host_template_id = gpsmap_templates.templateID WHERE id=".$did."");
				
				if($row['AP'] == 1){
					$fields_host_edit3['start'] = array(
					'friendly_name' => 'Starting Degree',
					'description' => 'Starting degree for directional area between 0-360',
					'method' => 'textbox',
					'max_length' => 4,
					'value' => '|arg1:start|',
					'default' => '0',
				);
				$fields_host_edit3['stop'] = array(
					'friendly_name' => 'Stopping Degree',
					'description' => 'Stopping degree for directional area, must be greater than the Starting Degree',
					'method' => 'textbox',
					'max_length' => 4,
					'value' => '|arg1:stop|',
					'default' => '360',
				);
				$fields_host_edit3['rdistance'] = array(
					'friendly_name' => 'Specify Radius',
					'description' => 'Manually specify radius for Access Point. Set to 0 to determine radius based on grouped devices',
					'method' => 'textbox',
					'max_length' => 10,
					'value' => '|arg1:rdistance|',
					'default' => '0',
				);
				}
			}
			$fields_host_edit3['groupnum'] = array(
				'friendly_name' => 'Group ID',
				'description' => 'Groups define what devices are included in the coverage overlay. Will be checked against AP device group number. 0 to disable',
				'method' => 'textbox',
				'max_length' => 3,
				'value' => '|arg1:groupnum|',
				'default' => '0',
			);
						
		} 
	}
	$fields_host_edit = $fields_host_edit3;
}

?>
